### PR TITLE
enable multiple pdf documents in one html page - fixing #5148

### PR DESCRIPTION
**Pull request by tmarrinan**

### CODE REVIEW

**Comment by yurydelendik** [comment]
There is document cleanup and destroy functionality. It removes all registered fonts. Does it still do it per document or all documents?

File: src/core/fonts.js [hunk 1]
```diff
@@ -2632,7 +2632,7 @@ var Font = (function FontClosure() {
   Font.getFontID = (function () {
     var ID = 1;
     return function Font_getFontID() {
-      return String(ID++);
+      return PDFJS.documentID + "_" + String(ID++);
```
Comment (timvandermeij) [inline src/core/fonts.js]: Nit: strings must use single quotes for lint to pass, i.e., change `"_"` to `'_'`

File: src/display/api.js [hunk 2]
```diff
@@ -178,6 +178,13 @@ PDFJS.openExternalLinksInNewWindow = (
     false : PDFJS.openExternalLinksInNewWindow);
 
 /**
+ * Unique identifier for each PDF document loaded in a single page
+ * @var {number}
+ */
+ PDFJS.documentID = (PDFJS.documentID === undefined ?
```
Comment (timvandermeij) [inline src/display/api.js]: Why would we need to expose this from the API? It seems like something we would want to keep internal.

**Comment by tmarrinan** [comment]
How would you propose not exposing `PDFJS.documentID` from the API - I don't think JavaScript allows private member variables? Or do you just mean assigning it an initial value of `0`, regardless of if it was previously defined?

**Comment by tmarrinan** [comment]
Where can I find the document cleanup and destroy functions?
Is there an example where a PDF is destroyed, so I can test to see if this affects all documents or just the one being destroyed?

**Comment by timvandermeij** [comment]
I would say that this should not be in the API at all because it is in internal way of working with multiple PDFs. An implementer working with the API should not be worried with these internal details. I think PDF.js should just append the internal document ID (or a random number perhaps) to the font as you do now, but just not expose it through the API.

**Comment by tmarrinan** [comment]
I agree that would be ideal, but how would this be accomplished in JavaScript? There are no private members, and so far I've been unable to trace the path between creating a document (where I create the documentID) and where font names are created. The documentID variable would have to be passed to each function in between (which also seems non-ideal, since the documentID is really a member of the document, not just a parameter to some function).

**Comment by tmarrinan** [comment]
I have now added the documentID initialization to the same place as the build and version numbers. Surely it is not intended that users modify these values, yet they are still treated as members of the PDFJS object. I envision the documentID the same way.

**Comment by tmarrinan** [comment]
I tried with random numbers - this does keep it completely unexposed. However, there is no longer a guarantee that font's won't clash. Two documents could create a font with the same id if the random number generator picks the same number.

File: src/pdf.js [hunk 3]
```diff
@@ -28,6 +28,7 @@ if (typeof PDFJS === 'undefined') {
 //#if BUNDLE_BUILD
 //#expand PDFJS.build = '__BUNDLE_BUILD__';
 //#endif
+//#expand PDFJS.documentID = 0;
```
Comment (yurydelendik) [inline src/pdf.js]: Define documentID as a variable in the api.js (it will be hidden by closure), not need to expose that in the PDFJS namespace.

**Comment by yurydelendik** [comment]
> Where can I find the document cleanup and destroy functions?

e.g. at https://github.com/mozilla/pdf.js/blob/master/src/display/api.js#L1368 which calls https://github.com/mozilla/pdf.js/blob/master/src/display/font_loader.js#L38

File: src/pdf.js (hunk 3 above)
Comment (tmarrinan) [inline src/pdf.js]: I initially created it in api.js. Where specifically can I put it so it is available in fonts.js, and worker.js as well?

**Comment by tmarrinan** [comment]
Thanks for pointing me to where in the code the delete happens.
How does one utilize this? Is there a simple example where I can load a PDF then delete / destroy it?

**Comment by tmarrinan** [comment]
I just traced the path from creating a document to generating fonts, and now successfully have the `documentID` hidden within closure so that it is not publicly accessible through the `PDFJS` object

**Comment by timvandermeij** [comment]
Could you squash the commits into one? See https://github.com/mozilla/pdf.js/wiki/Squashing-Commits on how to do that easily.

File: src/core/fonts.js [hunk 4]
```diff
@@ -2631,8 +2631,8 @@ var Font = (function FontClosure() {
 
   Font.getFontID = (function () {
     var ID = 1;
-    return function Font_getFontID() {
-      return String(ID++);
+    return function Font_getFontID(documentID) {
+      return documentID + '_' + String(ID++);
```
Comment (Rob--W) [inline src/core/fonts.js]: `String` can be removed since the value is implicitly converted to a string.

File: src/display/api.js [hunk 5]
```diff
@@ -177,6 +177,7 @@ PDFJS.openExternalLinksInNewWindow = (
   PDFJS.openExternalLinksInNewWindow === undefined ?
     false : PDFJS.openExternalLinksInNewWindow);
 
+
```
Comment (Rob--W) [inline src/display/api.js]: Why was this blank line added? If it was done by a mistake, please remove it.

**Comment by timvandermeij** [comment]
Blocked by #5992.

**Comment by yurydelendik** [comment]
See also #6477

**Comment by timvandermeij** [comment]
Superseded by #6571. Closing as that solution is more complete and this PR is stale.

**Pull request closed without merging**